// @pnp/spfx-controls-react/lib/controls/carousel/Carousel.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./Carousel.module.css');
var styles = {
    container: 'container_980abf6a',
    contentContainer: 'contentContainer_980abf6a',
    slideWrapper: 'slideWrapper_980abf6a',
    left: 'left_980abf6a',
    right: 'right_980abf6a',
    slideLeft: 'slideLeft_980abf6a',
    slideleft: 'slideleft_980abf6a',
    slideRight: 'slideRight_980abf6a',
    slideright: 'slideright_980abf6a',
    slideFromRight: 'slideFromRight_980abf6a',
    slidefromright: 'slidefromright_980abf6a',
    slideFromLeft: 'slideFromLeft_980abf6a',
    slidefromleft: 'slidefromleft_980abf6a',
    carouselImage: 'carouselImage_980abf6a',
    image: 'image_980abf6a',
    staticDetails: 'staticDetails_980abf6a',
    details: 'details_980abf6a',
    title: 'title_980abf6a',
    dynamicDetails: 'dynamicDetails_980abf6a',
    indicators: 'indicators_980abf6a',
    rectangle: 'rectangle_980abf6a',
    square: 'square_980abf6a',
    circle: 'circle_980abf6a',
    active: 'active_980abf6a',
    loadingComponent: 'loadingComponent_980abf6a',
    buttonLocations: 'buttonLocations_980abf6a',
    centralButtonsContainer: 'centralButtonsContainer_980abf6a',
    topButtonsContainer: 'topButtonsContainer_980abf6a',
    bottomButtonsContainer: 'bottomButtonsContainer_980abf6a',
    buttonsOnlyPrevButton: 'buttonsOnlyPrevButton_980abf6a',
    buttonsOnlyNextButton: 'buttonsOnlyNextButton_980abf6a',
    buttonsContainer: 'buttonsContainer_980abf6a',
    buttonsOnlyContainer: 'buttonsOnlyContainer_980abf6a',
    blockButtonsContainer: 'blockButtonsContainer_980abf6a',
    hiddenButtonsContainer: 'hiddenButtonsContainer_980abf6a',
};
exports.default = styles;
/* tslint:enable */
