// @pnp/spfx-controls-react/lib/controls/filePicker/LinkFilePickerTab/LinkFilePickerTab.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./LinkFilePickerTab.module.css');
var styles = {
    filePicker: 'filePicker_43d0bea1',
    tabsContainer: 'tabsContainer_43d0bea1',
    tabContainer: 'tabContainer_43d0bea1',
    nav: 'nav_43d0bea1',
    tabHeaderContainer: 'tabHeaderContainer_43d0bea1',
    breadcrumbNav: 'breadcrumbNav_43d0bea1',
    breadcrumbNavItem: 'breadcrumbNavItem_43d0bea1',
    tab: 'tab_43d0bea1',
    tabFiles: 'tabFiles_43d0bea1',
    scrollablePaneWrapper: 'scrollablePaneWrapper_43d0bea1',
    actionButtonsContainer: 'actionButtonsContainer_43d0bea1',
    actionButtons: 'actionButtons_43d0bea1',
    actionButton: 'actionButton_43d0bea1',
    tabHeader: 'tabHeader_43d0bea1',
    selectedTab: 'selectedTab_43d0bea1',
    header: 'header_43d0bea1',
    xlg: 'xlg_43d0bea1',
    xxlg: 'xxlg_43d0bea1',
    xxxlg: 'xxxlg_43d0bea1',
    xxxxlg: 'xxxxlg_43d0bea1',
    itemPickerTopBar: 'itemPickerTopBar_43d0bea1',
    commandBarNoChevron: 'commandBarNoChevron_43d0bea1',
    fileItem: 'fileItem_43d0bea1',
    folderItem: 'folderItem_43d0bea1',
    fileTypeIcon: 'fileTypeIcon_43d0bea1',
    fileTypeIconIcon: 'fileTypeIconIcon_43d0bea1',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_43d0bea1',
    iconColumnHeader: 'iconColumnHeader_43d0bea1',
    fileRow: 'fileRow_43d0bea1',
    folderRow: 'folderRow_43d0bea1',
    emptyFolder: 'emptyFolder_43d0bea1',
    emptyFolderImage: 'emptyFolderImage_43d0bea1',
    emptyFolderImageTag: 'emptyFolderImageTag_43d0bea1',
    emptyFolderTitle: 'emptyFolderTitle_43d0bea1',
    emptyFolderSubText: 'emptyFolderSubText_43d0bea1',
    emptyFolderMobile: 'emptyFolderMobile_43d0bea1',
    emptyFolderPc: 'emptyFolderPc_43d0bea1',
    emptyFolderIcon: 'emptyFolderIcon_43d0bea1',
    linkTextField: 'linkTextField_43d0bea1',
    localTabSinglePreview: 'localTabSinglePreview_43d0bea1',
    localTabSinglePreviewImage: 'localTabSinglePreviewImage_43d0bea1',
};
exports.default = styles;
/* tslint:enable */
