// @pnp/spfx-controls-react/lib/controls/filePicker/UploadFilePickerTab/UploadFilePickerTab.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./UploadFilePickerTab.module.css');
var styles = {
    filePicker: 'filePicker_59978c56',
    tabsContainer: 'tabsContainer_59978c56',
    tabContainer: 'tabContainer_59978c56',
    nav: 'nav_59978c56',
    tabHeaderContainer: 'tabHeaderContainer_59978c56',
    breadcrumbNav: 'breadcrumbNav_59978c56',
    breadcrumbNavItem: 'breadcrumbNavItem_59978c56',
    tab: 'tab_59978c56',
    tabFiles: 'tabFiles_59978c56',
    scrollablePaneWrapper: 'scrollablePaneWrapper_59978c56',
    actionButtonsContainer: 'actionButtonsContainer_59978c56',
    actionButtons: 'actionButtons_59978c56',
    actionButton: 'actionButton_59978c56',
    tabHeader: 'tabHeader_59978c56',
    selectedTab: 'selectedTab_59978c56',
    header: 'header_59978c56',
    xlg: 'xlg_59978c56',
    xxlg: 'xxlg_59978c56',
    xxxlg: 'xxxlg_59978c56',
    xxxxlg: 'xxxxlg_59978c56',
    itemPickerTopBar: 'itemPickerTopBar_59978c56',
    commandBarNoChevron: 'commandBarNoChevron_59978c56',
    fileItem: 'fileItem_59978c56',
    folderItem: 'folderItem_59978c56',
    fileTypeIcon: 'fileTypeIcon_59978c56',
    fileTypeIconIcon: 'fileTypeIconIcon_59978c56',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_59978c56',
    iconColumnHeader: 'iconColumnHeader_59978c56',
    fileRow: 'fileRow_59978c56',
    folderRow: 'folderRow_59978c56',
    emptyFolder: 'emptyFolder_59978c56',
    emptyFolderImage: 'emptyFolderImage_59978c56',
    emptyFolderImageTag: 'emptyFolderImageTag_59978c56',
    emptyFolderTitle: 'emptyFolderTitle_59978c56',
    emptyFolderSubText: 'emptyFolderSubText_59978c56',
    emptyFolderMobile: 'emptyFolderMobile_59978c56',
    emptyFolderPc: 'emptyFolderPc_59978c56',
    emptyFolderIcon: 'emptyFolderIcon_59978c56',
    localTabSinglePreview: 'localTabSinglePreview_59978c56',
    localTabSinglePreviewImage: 'localTabSinglePreviewImage_59978c56',
    localTabFilename: 'localTabFilename_59978c56',
    localTabLabel: 'localTabLabel_59978c56',
    localTabInput: 'localTabInput_59978c56',
};
exports.default = styles;
/* tslint:enable */
