// @pnp/spfx-controls-react/lib/controls/filePicker/OneDriveFilesTab/OneDriveFilesTab.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./OneDriveFilesTab.module.css');
var styles = {
    filePicker: 'filePicker_d53bb33c',
    tabsContainer: 'tabsContainer_d53bb33c',
    tabContainer: 'tabContainer_d53bb33c',
    nav: 'nav_d53bb33c',
    tabHeaderContainer: 'tabHeaderContainer_d53bb33c',
    breadcrumbNav: 'breadcrumbNav_d53bb33c',
    breadcrumbNavItem: 'breadcrumbNavItem_d53bb33c',
    tab: 'tab_d53bb33c',
    tabFiles: 'tabFiles_d53bb33c',
    scrollablePaneWrapper: 'scrollablePaneWrapper_d53bb33c',
    actionButtonsContainer: 'actionButtonsContainer_d53bb33c',
    actionButtons: 'actionButtons_d53bb33c',
    actionButton: 'actionButton_d53bb33c',
    tabHeader: 'tabHeader_d53bb33c',
    selectedTab: 'selectedTab_d53bb33c',
    header: 'header_d53bb33c',
    xlg: 'xlg_d53bb33c',
    xxlg: 'xxlg_d53bb33c',
    xxxlg: 'xxxlg_d53bb33c',
    xxxxlg: 'xxxxlg_d53bb33c',
    itemPickerTopBar: 'itemPickerTopBar_d53bb33c',
    commandBarNoChevron: 'commandBarNoChevron_d53bb33c',
    fileItem: 'fileItem_d53bb33c',
    folderItem: 'folderItem_d53bb33c',
    fileTypeIcon: 'fileTypeIcon_d53bb33c',
    fileTypeIconIcon: 'fileTypeIconIcon_d53bb33c',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_d53bb33c',
    iconColumnHeader: 'iconColumnHeader_d53bb33c',
    fileRow: 'fileRow_d53bb33c',
    folderRow: 'folderRow_d53bb33c',
    emptyFolder: 'emptyFolder_d53bb33c',
    emptyFolderImage: 'emptyFolderImage_d53bb33c',
    emptyFolderImageTag: 'emptyFolderImageTag_d53bb33c',
    emptyFolderTitle: 'emptyFolderTitle_d53bb33c',
    emptyFolderSubText: 'emptyFolderSubText_d53bb33c',
    emptyFolderMobile: 'emptyFolderMobile_d53bb33c',
    emptyFolderPc: 'emptyFolderPc_d53bb33c',
    emptyFolderIcon: 'emptyFolderIcon_d53bb33c',
};
exports.default = styles;
/* tslint:enable */
