// @microsoft/sp-application-base/lib/navigator/NavigationOrchestrator.js
import { Guid, Session, Validate, _SPEventManager } from '@microsoft/sp-core-library';
import { _Diagnostics, _LogSource, _TraceLogger } from '@microsoft/sp-diagnostics';
import { _ManifestStore, _SPLoaderFlights } from '@microsoft/sp-loader';
import { PageContext, SharePointPageContextDataProvider } from '@microsoft/sp-page-context';
import { _Telemetry, _PerformanceLogger } from '@ms/sp-telemetry';
import ApplicationManager from '../ApplicationManager';
import BaseApplication from '../BaseApplication';
import ApplicationCustomizerLoader from '../extensibility/ApplicationCustomizerLoader';
import SPPageChrome from '../pageChrome/SPPageChrome';
import SPThemeProvider from '../pageChrome/SPThemeProvider';
import { NavigationOperation, NavigationUnsupportedReason } from './INavigationResult';
import SPMasterChrome from '../pageChrome/SPMasterChrome';
import { Flights } from '../common/Flights';
import { Killswitches } from '../common/Killswitches';
var NavigationOrchestrator =  (function () {
    function NavigationOrchestrator(serviceScope, applicationManager, navigator) {
        this._numAppsInitializedInSession = 0;
        Validate.isNotNullOrUndefined(serviceScope, 'serviceScope');
        this._serviceScope = serviceScope;
        this._applicationManager = applicationManager || new ApplicationManager(serviceScope, navigator);
        this._pageContext = serviceScope.consume(PageContext.serviceKey);
        this._applicationCustomizerLoader = new ApplicationCustomizerLoader(serviceScope);
    }
    NavigationOrchestrator.raiseNavigatedEvent = function () {
        _SPEventManager.instance.raiseStickyEvent(BaseApplication._navigatedEventName, {});
    };
    NavigationOrchestrator.prototype.navigate = function (preloadedData) {
        var _this = this;
        var currentAppId = this._applicationManager.currentApplicationId;
        var appLoaded = !currentAppId.equals(Guid.empty);
        var isSameApp = currentAppId.equals(Guid.tryParse(preloadedData.clientSideApplicationId) || Guid.empty);
        if (Killswitches.isInplaceNavigationKSActive()) {
            if (!_SPLoaderFlights._useNewBootSequence()) {
                if (appLoaded && !isSameApp) {
                    return Promise.resolve({
                        preloadedData: preloadedData,
                        application: undefined,
                        operation: NavigationOperation.Unsupported
                    });
                }
            }
        }
        else {
            var isAppIdSameOrEmpty = isSameApp || !appLoaded;
            var unsupportedReason = this._isNavigationSupportedWithReason(preloadedData, this._pageContext, isAppIdSameOrEmpty);
            if (unsupportedReason !== undefined) {
                return Promise.resolve({
                    preloadedData: preloadedData,
                    application: undefined,
                    operation: NavigationOperation.Unsupported,
                    unsupportedReason: unsupportedReason
                });
            }
        }
        this._fixPreloadedDataItem(preloadedData);
        var isCrossSite = this._isCrossSite(preloadedData, this._pageContext);
        this._pageContext.initialize(SharePointPageContextDataProvider._createPageContextData(preloadedData.spPageContextInfo), preloadedData.spPageContextInfo);
        _ManifestStore.instance.replaceManifests(preloadedData.manifests);
        if (!ApplicationManager._isChromelessApplication(preloadedData.clientSideApplicationId)) {
            if (Flights.useNewChromeSequence() && !this._masterChrome) {
                this._masterChrome = new SPMasterChrome();
            }
            if (Flights.useNextGenSPA) {
                if (!this._pageChrome) {
                    this._pageChrome = new SPPageChrome(this._serviceScope, this._masterChrome);
                }
            }
            else {
                this._pageChrome = new SPPageChrome(this._serviceScope, this._masterChrome);
            }
        }
        if (isCrossSite && !ApplicationManager._isChromelessApplication(preloadedData.clientSideApplicationId)) {
            if (Killswitches.themeProviderOverrideKSActive()) {
                var themeProvider = new SPThemeProvider(this._serviceScope);
                themeProvider.loadThemedStyles();
            }
            else if (appLoaded && isSameApp) {
                this._applicationManager.currentApplication._loadTheme();
            }
        }
        this._updateTelemetry(preloadedData);
        preloadedData.spPageContextInfo = undefined;
        Session._changePage();
        NavigationOrchestrator.raiseNavigatedEvent();
        if (_SPLoaderFlights._useNewBootSequence()) {
            if (isSameApp) {
                this._loadApplicationCustomizers(preloadedData);
                return Promise.resolve({
                    preloadedData: preloadedData,
                    application: this._applicationManager.currentApplication,
                    operation: NavigationOperation.UpdateContext
                });
            }
            else {
                _PerformanceLogger.markCrossAppNavigationStart();
                return this._applicationManager.startApplication(preloadedData, this._pageChrome).then(function (application) {
                    _this._numAppsInitializedInSession++;
                    _this._loadApplicationCustomizers(preloadedData);
                    return Promise.resolve({
                        preloadedData: preloadedData,
                        application: application,
                        operation: NavigationOperation.LoadApplication
                    });
                });
            }
        }
        else {
            return Promise.resolve({
                preloadedData: preloadedData,
                application: undefined,
                operation: NavigationOperation.UpdateContext
            });
        }
    };
    NavigationOrchestrator.prototype._loadApplicationCustomizers = function (preloadedData) {
        if (preloadedData.clientSideApplicationId !== NavigationOrchestrator.classicPagesComponentId &&
            preloadedData.clientSideApplicationId !== NavigationOrchestrator.workbenchComponentId) {
            _TraceLogger.logVerbose(NavigationOrchestrator._logSource, 'Loading application customizers');
            return this._applicationCustomizerLoader.loadExtensions(preloadedData.customActions);
        }
        else {
            _TraceLogger.logVerbose(NavigationOrchestrator._logSource, 'Application Customizer is not supported for classic pages and workbench.');
            return Promise.resolve();
        }
    };
    NavigationOrchestrator.prototype._updateTelemetry = function (preloadedData) {
        var _a = this._pageContext.legacyPageContext, correlationId = _a.correlationId, siteId = _a.siteId, webId = _a.webId;
        var emptyIdValueForMerge = ''; 
        var listId = (this._pageContext.list && this._pageContext.list.id.toString()) ||
            emptyIdValueForMerge;
        var uniqueId = (preloadedData.item && preloadedData.item.UniqueId) ||
            emptyIdValueForMerge;
        var manifestAlias = preloadedData && preloadedData.manifests && preloadedData.manifests.length > 0 ?
            preloadedData.manifests[0].alias :
            '';
        _Telemetry.updateSettings({
            siteId: siteId,
            webId: webId,
            listId: listId,
            listItemUniqueId: uniqueId,
            correlationId: correlationId
        }, manifestAlias );
        _Diagnostics.updateSettings({
            siteId: siteId,
            webId: webId,
            listId: listId,
            listItemUniqueId: uniqueId,
            correlationId: correlationId
        });
    };
    NavigationOrchestrator.prototype._isCrossSite = function (preloadedData, pageContext) {
        if (!pageContext.isInitialized) { 
            return false;
        }
        return !Guid.parse(preloadedData.spPageContextInfo.webId).equals(pageContext.web.id) ||
            !Guid.parse(preloadedData.spPageContextInfo.siteId).equals(pageContext.site.id);
    };
    NavigationOrchestrator.prototype._fixPreloadedDataItem = function (preloadedData) {
        var INVALID_PAGE_ITEM_ID = -1;
        if (preloadedData.spPageContextInfo.pageItemId === INVALID_PAGE_ITEM_ID &&
            preloadedData.item &&
            preloadedData.item.ID !== INVALID_PAGE_ITEM_ID) {
            preloadedData.spPageContextInfo.pageItemId = preloadedData.item.ID;
        }
        return preloadedData;
    };
    NavigationOrchestrator.prototype._isNavigationSupportedWithReason = function (preloadedData, pageContext, isSameAppId) {
        if (!_SPLoaderFlights._useNewBootSequence()) {
            return isSameAppId ? undefined : NavigationUnsupportedReason.AppTransitionNotEnabled;
        }
        var isDifferentLocale = pageContext.isInitialized &&
            pageContext.cultureInfo.currentUICultureName !== preloadedData.spPageContextInfo.currentUICultureName;
        if (isDifferentLocale) {
            return NavigationUnsupportedReason.LocaleChange;
        }
        if (!isSameAppId && this._numAppsInitializedInSession > NavigationOrchestrator.maxSupportedAppsInSession) {
            return NavigationUnsupportedReason.TransitionsExceeded;
        }
        return undefined;
    };
    NavigationOrchestrator.classicPagesComponentId = 'eb4b666b-5c29-4dad-9a99-23613f21a2b7';
    NavigationOrchestrator.workbenchComponentId = '8be81a5c-af38-4bb2-af97-afa3b64dfbed';
    NavigationOrchestrator._logSource = _LogSource.create('NavigationOrchestrator');
    NavigationOrchestrator.maxSupportedAppsInSession = 7;
    return NavigationOrchestrator;
}());
export default NavigationOrchestrator;
