// @pnp/spfx-controls-react/lib/controls/filePicker/controls/FolderTile/FolderTile.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./FolderTile.module.css');
var styles = {
    filePicker: 'filePicker_c6701ef7',
    tabsContainer: 'tabsContainer_c6701ef7',
    tabContainer: 'tabContainer_c6701ef7',
    nav: 'nav_c6701ef7',
    tabHeaderContainer: 'tabHeaderContainer_c6701ef7',
    breadcrumbNav: 'breadcrumbNav_c6701ef7',
    breadcrumbNavItem: 'breadcrumbNavItem_c6701ef7',
    tab: 'tab_c6701ef7',
    tabFiles: 'tabFiles_c6701ef7',
    scrollablePaneWrapper: 'scrollablePaneWrapper_c6701ef7',
    actionButtonsContainer: 'actionButtonsContainer_c6701ef7',
    actionButtons: 'actionButtons_c6701ef7',
    actionButton: 'actionButton_c6701ef7',
    tabHeader: 'tabHeader_c6701ef7',
    selectedTab: 'selectedTab_c6701ef7',
    header: 'header_c6701ef7',
    xlg: 'xlg_c6701ef7',
    xxlg: 'xxlg_c6701ef7',
    xxxlg: 'xxxlg_c6701ef7',
    xxxxlg: 'xxxxlg_c6701ef7',
    itemPickerTopBar: 'itemPickerTopBar_c6701ef7',
    commandBarNoChevron: 'commandBarNoChevron_c6701ef7',
    fileItem: 'fileItem_c6701ef7',
    folderItem: 'folderItem_c6701ef7',
    fileTypeIcon: 'fileTypeIcon_c6701ef7',
    fileTypeIconIcon: 'fileTypeIconIcon_c6701ef7',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_c6701ef7',
    iconColumnHeader: 'iconColumnHeader_c6701ef7',
    fileRow: 'fileRow_c6701ef7',
    folderRow: 'folderRow_c6701ef7',
    emptyFolder: 'emptyFolder_c6701ef7',
    emptyFolderImage: 'emptyFolderImage_c6701ef7',
    emptyFolderImageTag: 'emptyFolderImageTag_c6701ef7',
    emptyFolderTitle: 'emptyFolderTitle_c6701ef7',
    emptyFolderSubText: 'emptyFolderSubText_c6701ef7',
    emptyFolderMobile: 'emptyFolderMobile_c6701ef7',
    emptyFolderPc: 'emptyFolderPc_c6701ef7',
    emptyFolderIcon: 'emptyFolderIcon_c6701ef7',
    aboveNameplate: 'aboveNameplate_c6701ef7',
    tileContent: 'tileContent_c6701ef7',
    foreground: 'foreground_c6701ef7',
    grid: 'grid_c6701ef7',
    cellContent: 'cellContent_c6701ef7',
    content: 'content_c6701ef7',
    tile: 'tile_c6701ef7',
    invokable: 'invokable_c6701ef7',
    namePlate: 'namePlate_c6701ef7',
    nameplate: 'nameplate_c6701ef7',
    name: 'name_c6701ef7',
    isLarge: 'isLarge_c6701ef7',
    activity: 'activity_c6701ef7',
    signalField: 'signalField_c6701ef7',
    compact: 'compact_c6701ef7',
    signalFieldValue: 'signalFieldValue_c6701ef7',
    folderList: 'folderList_c6701ef7',
    link: 'link_c6701ef7',
    cell: 'cell_c6701ef7',
    listCell: 'listCell_c6701ef7',
    label: 'label_c6701ef7',
    check: 'check_c6701ef7',
    selectable: 'selectable_c6701ef7',
    showCheck: 'showCheck_c6701ef7',
    selected: 'selected_c6701ef7',
    dialogMainOverride: 'dialogMainOverride_c6701ef7',
    itemTileBlankCover: 'itemTileBlankCover_c6701ef7',
    metadata: 'metadata_c6701ef7',
    isSmall: 'isSmall_c6701ef7',
    listPage: 'listPage_c6701ef7',
    odItemTile2FolderCover: 'odItemTile2FolderCover_c6701ef7',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_c6701ef7',
    folderCover: 'folderCover_c6701ef7',
    isSmall_88d29bd4: 'isSmall_88d29bd4_c6701ef7',
    folderCoverBack: 'folderCoverBack_c6701ef7',
    folderCoverFront: 'folderCoverFront_c6701ef7',
    folderCoverFrame: 'folderCoverFrame_c6701ef7',
    folderCoverContent: 'folderCoverContent_c6701ef7',
};
exports.default = styles;
/* tslint:enable */
