// @pnp/spfx-controls-react/lib/controls/filePicker/controls/DocumentTile/DocumentTile.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./DocumentTile.module.css');
var styles = {
    filePicker: 'filePicker_e49bec19',
    tabsContainer: 'tabsContainer_e49bec19',
    tabContainer: 'tabContainer_e49bec19',
    nav: 'nav_e49bec19',
    tabHeaderContainer: 'tabHeaderContainer_e49bec19',
    breadcrumbNav: 'breadcrumbNav_e49bec19',
    breadcrumbNavItem: 'breadcrumbNavItem_e49bec19',
    tab: 'tab_e49bec19',
    tabFiles: 'tabFiles_e49bec19',
    scrollablePaneWrapper: 'scrollablePaneWrapper_e49bec19',
    actionButtonsContainer: 'actionButtonsContainer_e49bec19',
    actionButtons: 'actionButtons_e49bec19',
    actionButton: 'actionButton_e49bec19',
    tabHeader: 'tabHeader_e49bec19',
    selectedTab: 'selectedTab_e49bec19',
    header: 'header_e49bec19',
    xlg: 'xlg_e49bec19',
    xxlg: 'xxlg_e49bec19',
    xxxlg: 'xxxlg_e49bec19',
    xxxxlg: 'xxxxlg_e49bec19',
    itemPickerTopBar: 'itemPickerTopBar_e49bec19',
    commandBarNoChevron: 'commandBarNoChevron_e49bec19',
    fileItem: 'fileItem_e49bec19',
    folderItem: 'folderItem_e49bec19',
    fileTypeIcon: 'fileTypeIcon_e49bec19',
    fileTypeIconIcon: 'fileTypeIconIcon_e49bec19',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_e49bec19',
    iconColumnHeader: 'iconColumnHeader_e49bec19',
    fileRow: 'fileRow_e49bec19',
    folderRow: 'folderRow_e49bec19',
    emptyFolder: 'emptyFolder_e49bec19',
    emptyFolderImage: 'emptyFolderImage_e49bec19',
    emptyFolderImageTag: 'emptyFolderImageTag_e49bec19',
    emptyFolderTitle: 'emptyFolderTitle_e49bec19',
    emptyFolderSubText: 'emptyFolderSubText_e49bec19',
    emptyFolderMobile: 'emptyFolderMobile_e49bec19',
    emptyFolderPc: 'emptyFolderPc_e49bec19',
    emptyFolderIcon: 'emptyFolderIcon_e49bec19',
    aboveNameplate: 'aboveNameplate_e49bec19',
    tileContent: 'tileContent_e49bec19',
    foreground: 'foreground_e49bec19',
    grid: 'grid_e49bec19',
    cellContent: 'cellContent_e49bec19',
    content: 'content_e49bec19',
    tile: 'tile_e49bec19',
    invokable: 'invokable_e49bec19',
    namePlate: 'namePlate_e49bec19',
    nameplate: 'nameplate_e49bec19',
    name: 'name_e49bec19',
    isLarge: 'isLarge_e49bec19',
    activity: 'activity_e49bec19',
    signalField: 'signalField_e49bec19',
    compact: 'compact_e49bec19',
    signalFieldValue: 'signalFieldValue_e49bec19',
    folderList: 'folderList_e49bec19',
    link: 'link_e49bec19',
    cell: 'cell_e49bec19',
    listCell: 'listCell_e49bec19',
    label: 'label_e49bec19',
    check: 'check_e49bec19',
    selectable: 'selectable_e49bec19',
    showCheck: 'showCheck_e49bec19',
    selected: 'selected_e49bec19',
    dialogMainOverride: 'dialogMainOverride_e49bec19',
    itemTileBlankCover: 'itemTileBlankCover_e49bec19',
    metadata: 'metadata_e49bec19',
    isSmall: 'isSmall_e49bec19',
    listPage: 'listPage_e49bec19',
    odItemTile2FolderCover: 'odItemTile2FolderCover_e49bec19',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_e49bec19',
    odItemTile2Image: 'odItemTile2Image_e49bec19',
    odImageFrame2: 'odImageFrame2_e49bec19',
    odImageFrame2Image: 'odImageFrame2Image_e49bec19',
    odImageFrame: 'odImageFrame_e49bec19',
    odImageStack: 'odImageStack_e49bec19',
    odImageStackTile: 'odImageStackTile_e49bec19',
    odImageTile: 'odImageTile_e49bec19',
    odImageTileBackground: 'odImageTileBackground_e49bec19',
    odImageTileImage: 'odImageTileImage_e49bec19',
    odItemTile2SmallIcon: 'odItemTile2SmallIcon_e49bec19',
};
exports.default = styles;
/* tslint:enable */
