// @pnp/spfx-controls-react/lib/controls/filePicker/controls/TilesList/TilesList.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./TilesList.module.css');
var styles = {
    filePicker: 'filePicker_60d7941d',
    tabsContainer: 'tabsContainer_60d7941d',
    tabContainer: 'tabContainer_60d7941d',
    nav: 'nav_60d7941d',
    tabHeaderContainer: 'tabHeaderContainer_60d7941d',
    breadcrumbNav: 'breadcrumbNav_60d7941d',
    breadcrumbNavItem: 'breadcrumbNavItem_60d7941d',
    tab: 'tab_60d7941d',
    tabFiles: 'tabFiles_60d7941d',
    scrollablePaneWrapper: 'scrollablePaneWrapper_60d7941d',
    actionButtonsContainer: 'actionButtonsContainer_60d7941d',
    actionButtons: 'actionButtons_60d7941d',
    actionButton: 'actionButton_60d7941d',
    tabHeader: 'tabHeader_60d7941d',
    selectedTab: 'selectedTab_60d7941d',
    header: 'header_60d7941d',
    xlg: 'xlg_60d7941d',
    xxlg: 'xxlg_60d7941d',
    xxxlg: 'xxxlg_60d7941d',
    xxxxlg: 'xxxxlg_60d7941d',
    itemPickerTopBar: 'itemPickerTopBar_60d7941d',
    commandBarNoChevron: 'commandBarNoChevron_60d7941d',
    fileItem: 'fileItem_60d7941d',
    folderItem: 'folderItem_60d7941d',
    fileTypeIcon: 'fileTypeIcon_60d7941d',
    fileTypeIconIcon: 'fileTypeIconIcon_60d7941d',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_60d7941d',
    iconColumnHeader: 'iconColumnHeader_60d7941d',
    fileRow: 'fileRow_60d7941d',
    folderRow: 'folderRow_60d7941d',
    emptyFolder: 'emptyFolder_60d7941d',
    emptyFolderImage: 'emptyFolderImage_60d7941d',
    emptyFolderImageTag: 'emptyFolderImageTag_60d7941d',
    emptyFolderTitle: 'emptyFolderTitle_60d7941d',
    emptyFolderSubText: 'emptyFolderSubText_60d7941d',
    emptyFolderMobile: 'emptyFolderMobile_60d7941d',
    emptyFolderPc: 'emptyFolderPc_60d7941d',
    emptyFolderIcon: 'emptyFolderIcon_60d7941d',
    aboveNameplate: 'aboveNameplate_60d7941d',
    tileContent: 'tileContent_60d7941d',
    foreground: 'foreground_60d7941d',
    grid: 'grid_60d7941d',
    cellContent: 'cellContent_60d7941d',
    content: 'content_60d7941d',
    tile: 'tile_60d7941d',
    invokable: 'invokable_60d7941d',
    namePlate: 'namePlate_60d7941d',
    nameplate: 'nameplate_60d7941d',
    name: 'name_60d7941d',
    isLarge: 'isLarge_60d7941d',
    activity: 'activity_60d7941d',
    signalField: 'signalField_60d7941d',
    compact: 'compact_60d7941d',
    signalFieldValue: 'signalFieldValue_60d7941d',
    folderList: 'folderList_60d7941d',
    link: 'link_60d7941d',
    cell: 'cell_60d7941d',
    listCell: 'listCell_60d7941d',
    label: 'label_60d7941d',
    check: 'check_60d7941d',
    selectable: 'selectable_60d7941d',
    showCheck: 'showCheck_60d7941d',
    selected: 'selected_60d7941d',
    dialogMainOverride: 'dialogMainOverride_60d7941d',
    itemTileBlankCover: 'itemTileBlankCover_60d7941d',
    metadata: 'metadata_60d7941d',
    isSmall: 'isSmall_60d7941d',
    listPage: 'listPage_60d7941d',
    odItemTile2FolderCover: 'odItemTile2FolderCover_60d7941d',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_60d7941d',
};
exports.default = styles;
/* tslint:enable */
