// @pnp/spfx-controls-react/lib/controls/filePicker/WebSearchTab/WebSearchTab.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./WebSearchTab.module.css');
var styles = {
    filePicker: 'filePicker_d0d43970',
    tabsContainer: 'tabsContainer_d0d43970',
    tabContainer: 'tabContainer_d0d43970',
    nav: 'nav_d0d43970',
    tabHeaderContainer: 'tabHeaderContainer_d0d43970',
    breadcrumbNav: 'breadcrumbNav_d0d43970',
    breadcrumbNavItem: 'breadcrumbNavItem_d0d43970',
    tab: 'tab_d0d43970',
    tabFiles: 'tabFiles_d0d43970',
    scrollablePaneWrapper: 'scrollablePaneWrapper_d0d43970',
    actionButtonsContainer: 'actionButtonsContainer_d0d43970',
    actionButtons: 'actionButtons_d0d43970',
    actionButton: 'actionButton_d0d43970',
    tabHeader: 'tabHeader_d0d43970',
    selectedTab: 'selectedTab_d0d43970',
    header: 'header_d0d43970',
    xlg: 'xlg_d0d43970',
    xxlg: 'xxlg_d0d43970',
    xxxlg: 'xxxlg_d0d43970',
    xxxxlg: 'xxxxlg_d0d43970',
    itemPickerTopBar: 'itemPickerTopBar_d0d43970',
    commandBarNoChevron: 'commandBarNoChevron_d0d43970',
    fileItem: 'fileItem_d0d43970',
    folderItem: 'folderItem_d0d43970',
    fileTypeIcon: 'fileTypeIcon_d0d43970',
    fileTypeIconIcon: 'fileTypeIconIcon_d0d43970',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_d0d43970',
    iconColumnHeader: 'iconColumnHeader_d0d43970',
    fileRow: 'fileRow_d0d43970',
    folderRow: 'folderRow_d0d43970',
    emptyFolder: 'emptyFolder_d0d43970',
    emptyFolderImage: 'emptyFolderImage_d0d43970',
    emptyFolderImageTag: 'emptyFolderImageTag_d0d43970',
    emptyFolderTitle: 'emptyFolderTitle_d0d43970',
    emptyFolderSubText: 'emptyFolderSubText_d0d43970',
    emptyFolderMobile: 'emptyFolderMobile_d0d43970',
    emptyFolderPc: 'emptyFolderPc_d0d43970',
    emptyFolderIcon: 'emptyFolderIcon_d0d43970',
    searchBoxContainer: 'searchBoxContainer_d0d43970',
    searchBoxMedium: 'searchBoxMedium_d0d43970',
    searchBox: 'searchBox_d0d43970',
    dropdownContainer: 'dropdownContainer_d0d43970',
    filterDropdown: 'filterDropdown_d0d43970',
    filePickerFolderCardGrid: 'filePickerFolderCardGrid_d0d43970',
    filePickerFolderCardTile: 'filePickerFolderCardTile_d0d43970',
    filePickerFolderCardPadder: 'filePickerFolderCardPadder_d0d43970',
    filePickerFolderCardSizer: 'filePickerFolderCardSizer_d0d43970',
    filePickerFolderCardImage: 'filePickerFolderCardImage_d0d43970',
    filePickerFolderCardLabel: 'filePickerFolderCardLabel_d0d43970',
    bingGridListCell: 'bingGridListCell_d0d43970',
    bingTile: 'bingTile_d0d43970',
    bingTileContent: 'bingTileContent_d0d43970',
    bingTileFrame: 'bingTileFrame_d0d43970',
    bingTileCheckCircle: 'bingTileCheckCircle_d0d43970',
    bingTileNamePlate: 'bingTileNamePlate_d0d43970',
    isSelected: 'isSelected_d0d43970',
    copyrightLabel: 'copyrightLabel_d0d43970',
    noResultLabel: 'noResultLabel_d0d43970',
    bingTileThumbnail: 'bingTileThumbnail_d0d43970',
    bingGrildList: 'bingGrildList_d0d43970',
};
exports.default = styles;
/* tslint:enable */
