// @pnp/spfx-controls-react/lib/controls/filePicker/StockImagesTab/StockImages.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./StockImages.module.css');
var styles = {
    filePicker: 'filePicker_f8391f5c',
    tabsContainer: 'tabsContainer_f8391f5c',
    tabContainer: 'tabContainer_f8391f5c',
    nav: 'nav_f8391f5c',
    tabHeaderContainer: 'tabHeaderContainer_f8391f5c',
    breadcrumbNav: 'breadcrumbNav_f8391f5c',
    breadcrumbNavItem: 'breadcrumbNavItem_f8391f5c',
    tab: 'tab_f8391f5c',
    tabFiles: 'tabFiles_f8391f5c',
    scrollablePaneWrapper: 'scrollablePaneWrapper_f8391f5c',
    actionButtonsContainer: 'actionButtonsContainer_f8391f5c',
    actionButtons: 'actionButtons_f8391f5c',
    actionButton: 'actionButton_f8391f5c',
    tabHeader: 'tabHeader_f8391f5c',
    selectedTab: 'selectedTab_f8391f5c',
    header: 'header_f8391f5c',
    xlg: 'xlg_f8391f5c',
    xxlg: 'xxlg_f8391f5c',
    xxxlg: 'xxxlg_f8391f5c',
    xxxxlg: 'xxxxlg_f8391f5c',
    itemPickerTopBar: 'itemPickerTopBar_f8391f5c',
    commandBarNoChevron: 'commandBarNoChevron_f8391f5c',
    fileItem: 'fileItem_f8391f5c',
    folderItem: 'folderItem_f8391f5c',
    fileTypeIcon: 'fileTypeIcon_f8391f5c',
    fileTypeIconIcon: 'fileTypeIconIcon_f8391f5c',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_f8391f5c',
    iconColumnHeader: 'iconColumnHeader_f8391f5c',
    fileRow: 'fileRow_f8391f5c',
    folderRow: 'folderRow_f8391f5c',
    emptyFolder: 'emptyFolder_f8391f5c',
    emptyFolderImage: 'emptyFolderImage_f8391f5c',
    emptyFolderImageTag: 'emptyFolderImageTag_f8391f5c',
    emptyFolderTitle: 'emptyFolderTitle_f8391f5c',
    emptyFolderSubText: 'emptyFolderSubText_f8391f5c',
    emptyFolderMobile: 'emptyFolderMobile_f8391f5c',
    emptyFolderPc: 'emptyFolderPc_f8391f5c',
    emptyFolderIcon: 'emptyFolderIcon_f8391f5c',
    stockImagesPickerContainer: 'stockImagesPickerContainer_f8391f5c',
    stockImagesPicker: 'stockImagesPicker_f8391f5c',
};
exports.default = styles;
/* tslint:enable */
