// @pnp/spfx-controls-react/lib/controls/filePicker/controls/FileBrowser/FileBrowser.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./FileBrowser.module.css');
var styles = {
    filePicker: 'filePicker_698f5438',
    tabsContainer: 'tabsContainer_698f5438',
    tabContainer: 'tabContainer_698f5438',
    nav: 'nav_698f5438',
    tabHeaderContainer: 'tabHeaderContainer_698f5438',
    breadcrumbNav: 'breadcrumbNav_698f5438',
    breadcrumbNavItem: 'breadcrumbNavItem_698f5438',
    tab: 'tab_698f5438',
    tabFiles: 'tabFiles_698f5438',
    scrollablePaneWrapper: 'scrollablePaneWrapper_698f5438',
    actionButtonsContainer: 'actionButtonsContainer_698f5438',
    actionButtons: 'actionButtons_698f5438',
    actionButton: 'actionButton_698f5438',
    tabHeader: 'tabHeader_698f5438',
    selectedTab: 'selectedTab_698f5438',
    header: 'header_698f5438',
    xlg: 'xlg_698f5438',
    xxlg: 'xxlg_698f5438',
    xxxlg: 'xxxlg_698f5438',
    xxxxlg: 'xxxxlg_698f5438',
    itemPickerTopBar: 'itemPickerTopBar_698f5438',
    commandBarNoChevron: 'commandBarNoChevron_698f5438',
    fileItem: 'fileItem_698f5438',
    folderItem: 'folderItem_698f5438',
    fileTypeIcon: 'fileTypeIcon_698f5438',
    fileTypeIconIcon: 'fileTypeIconIcon_698f5438',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_698f5438',
    iconColumnHeader: 'iconColumnHeader_698f5438',
    fileRow: 'fileRow_698f5438',
    folderRow: 'folderRow_698f5438',
    emptyFolder: 'emptyFolder_698f5438',
    emptyFolderImage: 'emptyFolderImage_698f5438',
    emptyFolderImageTag: 'emptyFolderImageTag_698f5438',
    emptyFolderTitle: 'emptyFolderTitle_698f5438',
    emptyFolderSubText: 'emptyFolderSubText_698f5438',
    emptyFolderMobile: 'emptyFolderMobile_698f5438',
    emptyFolderPc: 'emptyFolderPc_698f5438',
    emptyFolderIcon: 'emptyFolderIcon_698f5438',
};
exports.default = styles;
/* tslint:enable */
