// @pnp/spfx-controls-react/lib/controls/filePicker/SiteFilePickerTab/SiteFilePickerTab.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./SiteFilePickerTab.module.css');
var styles = {
    filePicker: 'filePicker_7687710d',
    tabsContainer: 'tabsContainer_7687710d',
    tabContainer: 'tabContainer_7687710d',
    nav: 'nav_7687710d',
    tabHeaderContainer: 'tabHeaderContainer_7687710d',
    breadcrumbNav: 'breadcrumbNav_7687710d',
    breadcrumbNavItem: 'breadcrumbNavItem_7687710d',
    tab: 'tab_7687710d',
    tabFiles: 'tabFiles_7687710d',
    scrollablePaneWrapper: 'scrollablePaneWrapper_7687710d',
    actionButtonsContainer: 'actionButtonsContainer_7687710d',
    actionButtons: 'actionButtons_7687710d',
    actionButton: 'actionButton_7687710d',
    tabHeader: 'tabHeader_7687710d',
    selectedTab: 'selectedTab_7687710d',
    header: 'header_7687710d',
    xlg: 'xlg_7687710d',
    xxlg: 'xxlg_7687710d',
    xxxlg: 'xxxlg_7687710d',
    xxxxlg: 'xxxxlg_7687710d',
    itemPickerTopBar: 'itemPickerTopBar_7687710d',
    commandBarNoChevron: 'commandBarNoChevron_7687710d',
    fileItem: 'fileItem_7687710d',
    folderItem: 'folderItem_7687710d',
    fileTypeIcon: 'fileTypeIcon_7687710d',
    fileTypeIconIcon: 'fileTypeIconIcon_7687710d',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_7687710d',
    iconColumnHeader: 'iconColumnHeader_7687710d',
    fileRow: 'fileRow_7687710d',
    folderRow: 'folderRow_7687710d',
    emptyFolder: 'emptyFolder_7687710d',
    emptyFolderImage: 'emptyFolderImage_7687710d',
    emptyFolderImageTag: 'emptyFolderImageTag_7687710d',
    emptyFolderTitle: 'emptyFolderTitle_7687710d',
    emptyFolderSubText: 'emptyFolderSubText_7687710d',
    emptyFolderMobile: 'emptyFolderMobile_7687710d',
    emptyFolderPc: 'emptyFolderPc_7687710d',
    emptyFolderIcon: 'emptyFolderIcon_7687710d',
};
exports.default = styles;
/* tslint:enable */
