// @pnp/spfx-controls-react/lib/controls/filePicker/FilePicker.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./FilePicker.module.css');
var styles = {
    filePicker: 'filePicker_371d2e9c',
    tabsContainer: 'tabsContainer_371d2e9c',
    tabContainer: 'tabContainer_371d2e9c',
    nav: 'nav_371d2e9c',
    tabHeaderContainer: 'tabHeaderContainer_371d2e9c',
    breadcrumbNav: 'breadcrumbNav_371d2e9c',
    breadcrumbNavItem: 'breadcrumbNavItem_371d2e9c',
    tab: 'tab_371d2e9c',
    tabFiles: 'tabFiles_371d2e9c',
    scrollablePaneWrapper: 'scrollablePaneWrapper_371d2e9c',
    actionButtonsContainer: 'actionButtonsContainer_371d2e9c',
    actionButtons: 'actionButtons_371d2e9c',
    actionButton: 'actionButton_371d2e9c',
    tabHeader: 'tabHeader_371d2e9c',
    selectedTab: 'selectedTab_371d2e9c',
    header: 'header_371d2e9c',
    xlg: 'xlg_371d2e9c',
    xxlg: 'xxlg_371d2e9c',
    xxxlg: 'xxxlg_371d2e9c',
    xxxxlg: 'xxxxlg_371d2e9c',
    itemPickerTopBar: 'itemPickerTopBar_371d2e9c',
    commandBarNoChevron: 'commandBarNoChevron_371d2e9c',
    fileItem: 'fileItem_371d2e9c',
    folderItem: 'folderItem_371d2e9c',
    fileTypeIcon: 'fileTypeIcon_371d2e9c',
    fileTypeIconIcon: 'fileTypeIconIcon_371d2e9c',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_371d2e9c',
    iconColumnHeader: 'iconColumnHeader_371d2e9c',
    fileRow: 'fileRow_371d2e9c',
    folderRow: 'folderRow_371d2e9c',
    emptyFolder: 'emptyFolder_371d2e9c',
    emptyFolderImage: 'emptyFolderImage_371d2e9c',
    emptyFolderImageTag: 'emptyFolderImageTag_371d2e9c',
    emptyFolderTitle: 'emptyFolderTitle_371d2e9c',
    emptyFolderSubText: 'emptyFolderSubText_371d2e9c',
    emptyFolderMobile: 'emptyFolderMobile_371d2e9c',
    emptyFolderPc: 'emptyFolderPc_371d2e9c',
    emptyFolderIcon: 'emptyFolderIcon_371d2e9c',
};
exports.default = styles;
/* tslint:enable */
