// @pnp/spfx-controls-react/lib/controls/filePicker/FilePicker.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Button_1 = require("office-ui-fabric-react/lib/components/Button");
var Panel_1 = require("office-ui-fabric-react/lib/components/Panel");
var Label_1 = require("office-ui-fabric-react/lib/components/Label");
var Nav_1 = require("office-ui-fabric-react/lib/Nav");
var css_1 = require("@uifabric/utilities/lib/css");
// Localization
var strings = require("ControlStrings");
var LinkFilePickerTab_1 = require("./LinkFilePickerTab/LinkFilePickerTab");
var UploadFilePickerTab_1 = require("./UploadFilePickerTab/UploadFilePickerTab");
var SiteFilePickerTab_1 = require("./SiteFilePickerTab/SiteFilePickerTab");
var WebSearchTab_1 = require("./WebSearchTab/WebSearchTab");
var RecentFilesTab_1 = require("./RecentFilesTab/RecentFilesTab");
var FilePicker_module_scss_1 = require("./FilePicker.module.scss");
var FileBrowserService_1 = require("../../services/FileBrowserService");
var OneDriveFilesTab_1 = require("./OneDriveFilesTab");
var OneDriveService_1 = require("../../services/OneDriveService");
var OrgAssetsService_1 = require("../../services/OrgAssetsService");
var FilesSearchService_1 = require("../../services/FilesSearchService");
var telemetry = require("../../common/telemetry");
var StockImages_1 = require("./StockImagesTab/StockImages");
var FilePicker = (function (_super) {
    __extends(FilePicker, _super);
    function FilePicker(props) {
        var _this = _super.call(this, props) || this;
        /**
         * Renders the panel header
         */
        _this._renderHeader = function () {
            return React.createElement("div", { className: "ms-Panel-header" },
                React.createElement("p", { className: css_1.css("ms-Panel-headerText", FilePicker_module_scss_1.default.header), role: "heading" }, strings.FilePickerHeader));
        };
        /**
         * Open the panel
         */
        _this._handleOpenPanel = function () {
            _this.setState({
                panelOpen: true,
                selectedTab: _this.getDefaultSelectedTabKey(_this.props, _this.state.organisationAssetsEnabled)
            });
        };
        /**
         * Closes the panel
         */
        _this._handleClosePanel = function () {
            _this.setState({
                panelOpen: false
            });
        };
        /**
         * On save action
         */
        _this._handleSave = function (filePickerResult) {
            _this.props.onSave(filePickerResult);
            _this.setState({
                panelOpen: false
            });
        };
        /**
         * Changes the selected tab when a link is selected
         */
        _this._handleLinkClick = function (ev, item) {
            _this.setState({ selectedTab: item.key });
        };
        /**
         * Prepares navigation panel options
         */
        _this._getNavPanelOptions = function () {
            var addUrl = _this.props.storeLastActiveTab !== false;
            var links = [];
            if (!_this.props.hideRecentTab) {
                links.push({
                    name: strings.RecentLinkLabel,
                    url: addUrl ? '#recent' : undefined,
                    icon: 'Recent',
                    key: 'keyRecent',
                });
            }
            if (!_this.props.hideStockImages) {
                links.push({
                    name: strings.StockImagesLinkLabel,
                    url: addUrl ? '#stockImages' : undefined,
                    key: 'keyStockImages',
                    icon: 'ImageSearch',
                });
            }
            if (_this.props.bingAPIKey && !_this.props.hideWebSearchTab) {
                links.push({
                    name: strings.WebSearchLinkLabel,
                    url: addUrl ? '#search' : undefined,
                    key: 'keyWeb',
                    icon: 'Search',
                });
            }
            if (!_this.props.hideOrganisationalAssetTab && _this.state.organisationAssetsEnabled) {
                links.push({
                    name: 'Your organisation',
                    url: addUrl ? '#orgAssets' : undefined,
                    icon: 'FabricFolderConfirm',
                    key: 'keyOrgAssets',
                });
            }
            if (!_this.props.hideOneDriveTab) {
                links.push({
                    name: "OneDrive",
                    url: addUrl ? '#onedrive' : undefined,
                    key: 'keyOneDrive',
                    icon: 'OneDrive',
                });
            }
            if (!_this.props.hideSiteFilesTab) {
                links.push({
                    name: strings.SiteLinkLabel,
                    url: addUrl ? '#globe' : undefined,
                    key: 'keySite',
                    icon: 'Globe',
                });
            }
            if (!_this.props.hideLocalUploadTab) {
                links.push({
                    name: strings.UploadLinkLabel,
                    url: addUrl ? '#upload' : undefined,
                    key: 'keyUpload',
                    icon: 'System'
                });
            }
            if (!_this.props.hideLinkUploadTab) {
                links.push({
                    name: strings.FromLinkLinkLabel,
                    url: addUrl ? '#link' : undefined,
                    key: 'keyLink',
                    icon: 'Link'
                });
            }
            var groups = [{ links: links }];
            return groups;
        };
        _this.getDefaultSelectedTabKey = function (props, orgAssetsEnabled) {
            if (!props.hideRecentTab) {
                return 'keyRecent';
            }
            if (!props.hideStockImages) {
                return 'keyStockImages';
            }
            if (props.bingAPIKey && !props.hideWebSearchTab) {
                return 'keyWeb';
            }
            if (!props.hideOrganisationalAssetTab && orgAssetsEnabled) {
                return 'keyOrgAssets';
            }
            if (!props.hideOneDriveTab) {
                return 'keyOneDrive';
            }
            if (!props.hideSiteFilesTab) {
                return 'keySite';
            }
            if (!props.hideLocalUploadTab) {
                return 'keyUpload';
            }
            if (!props.hideLinkUploadTab) {
                return 'keyLink';
            }
        };
        telemetry.track('ReactFilePicker', {});
        // Initialize file browser services
        _this.fileBrowserService = new FileBrowserService_1.FileBrowserService(props.context, _this.props.itemsCountQueryLimit);
        _this.oneDriveService = new OneDriveService_1.OneDriveService(props.context, _this.props.itemsCountQueryLimit);
        _this.orgAssetsService = new OrgAssetsService_1.OrgAssetsService(props.context, _this.props.itemsCountQueryLimit);
        _this.fileSearchService = new FilesSearchService_1.FilesSearchService(props.context, _this.props.bingAPIKey);
        _this.state = {
            panelOpen: false,
            organisationAssetsEnabled: false,
            showFullNav: true
        };
        return _this;
    }
    FilePicker.prototype.componentDidMount = function () {
        return __awaiter(this, void 0, void 0, function () {
            var orgAssetsEnabled, orgAssetsLibraries;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        orgAssetsEnabled = false;
                        if (!!this.props.hideOrganisationalAssetTab) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.orgAssetsService.getSiteMediaLibraries()];
                    case 1:
                        orgAssetsLibraries = _a.sent();
                        orgAssetsEnabled = orgAssetsLibraries ? true : false;
                        _a.label = 2;
                    case 2:
                        this.setState({
                            organisationAssetsEnabled: orgAssetsEnabled,
                            selectedTab: this.getDefaultSelectedTabKey(this.props, orgAssetsEnabled)
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    FilePicker.prototype.render = function () {
        var _this = this;
        // If no acceptable file type was passed, and we're expecting images, set the default image filter
        var accepts = this.props.accepts;
        var buttonClassName = this.props.buttonClassName ? this.props.buttonClassName : '';
        var panelClassName = this.props.panelClassName ? this.props.panelClassName : '';
        var linkTabProps = {
            accepts: accepts,
            context: this.props.context,
            onClose: function () { return _this._handleClosePanel(); },
            onSave: function (value) { _this._handleSave(value); }
        };
        var buttonProps = {
            text: this.props.buttonLabel,
            disabled: this.props.disabled,
            onClick: this._handleOpenPanel,
            className: "pnp__file-picker__button " + buttonClassName
        };
        return (React.createElement("div", { className: "pnp__file-picker" },
            this.props.label && React.createElement(Label_1.Label, { required: this.props.required }, this.props.label),
            this.props.buttonIcon ?
                React.createElement(Button_1.ActionButton, __assign({ iconProps: { iconName: this.props.buttonIcon } }, buttonProps)) :
                React.createElement(Button_1.PrimaryButton, __assign({}, buttonProps)),
            React.createElement(Panel_1.Panel, { isOpen: this.state.panelOpen, isBlocking: true, hasCloseButton: true, className: "pnp__file-picker__panel " + FilePicker_module_scss_1.default.filePicker + " " + panelClassName, onDismiss: this._handleClosePanel, type: Panel_1.PanelType.extraLarge, isFooterAtBottom: true, onRenderNavigation: function () { return undefined; }, headerText: strings.FilePickerHeader, isLightDismiss: true, onRenderHeader: function () { return _this._renderHeader(); } },
                React.createElement("div", { className: FilePicker_module_scss_1.default.nav },
                    React.createElement(Nav_1.Nav, { groups: this._getNavPanelOptions(), selectedKey: this.state.selectedTab, onLinkClick: function (ev, item) { return _this._handleLinkClick(ev, item); } })),
                React.createElement("div", { className: FilePicker_module_scss_1.default.tabsContainer },
                    this.state.selectedTab === "keyLink" &&
                        React.createElement(LinkFilePickerTab_1.default, __assign({ fileSearchService: this.fileSearchService, allowExternalTenantLinks: true }, linkTabProps)),
                    this.state.selectedTab === "keyUpload" &&
                        React.createElement(UploadFilePickerTab_1.default, __assign({}, linkTabProps)),
                    this.state.selectedTab === "keySite" &&
                        React.createElement(SiteFilePickerTab_1.default, __assign({ fileBrowserService: this.fileBrowserService }, linkTabProps)),
                    this.state.selectedTab === "keyOrgAssets" &&
                        React.createElement(SiteFilePickerTab_1.default, __assign({ breadcrumbFirstNode: {
                                isCurrentItem: true,
                                text: strings.OrgAssetsTabLabel,
                                key: "keyOrgAssets"
                            }, fileBrowserService: this.orgAssetsService }, linkTabProps)),
                    this.state.selectedTab === "keyWeb" &&
                        React.createElement(WebSearchTab_1.default, __assign({ bingSearchService: this.fileSearchService }, linkTabProps)),
                    this.state.selectedTab === "keyOneDrive" &&
                        React.createElement(OneDriveFilesTab_1.OneDriveFilesTab, __assign({ oneDriveService: this.oneDriveService }, linkTabProps)),
                    this.state.selectedTab === "keyRecent" &&
                        React.createElement(RecentFilesTab_1.default, __assign({ fileSearchService: this.fileSearchService }, linkTabProps)),
                    this.state.selectedTab === "keyStockImages" &&
                        React.createElement(StockImages_1.StockImages, __assign({ language: this.props.context.pageContext.cultureInfo.currentCultureName, fileSearchService: this.fileSearchService }, linkTabProps))))));
    };
    return FilePicker;
}(React.Component));
exports.FilePicker = FilePicker;
