// @pnp/spfx-controls-react/lib/controls/taxonomyPicker/TaxonomyPicker.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./TaxonomyPicker.module.css');
var styles = {
    contextualMenu: 'contextualMenu_bbbc76e9',
    listItem: 'listItem_bbbc76e9',
    termField: 'termField_bbbc76e9',
    termFieldInput: 'termFieldInput_bbbc76e9',
    termFieldButton: 'termFieldButton_bbbc76e9',
    termset: 'termset_bbbc76e9',
    termSetSelectable: 'termSetSelectable_bbbc76e9',
    termSetSelector: 'termSetSelector_bbbc76e9',
    term: 'term_bbbc76e9',
    termEnabled: 'termEnabled_bbbc76e9',
    termDisabled: 'termDisabled_bbbc76e9',
    termNoTagging: 'termNoTagging_bbbc76e9',
    actions: 'actions_bbbc76e9',
    termBasePicker: 'termBasePicker_bbbc76e9',
    termSuggestion: 'termSuggestion_bbbc76e9',
    termSuggestionSubTitle: 'termSuggestionSubTitle_bbbc76e9',
    pickedTermRoot: 'pickedTermRoot_bbbc76e9',
    pickedTermText: 'pickedTermText_bbbc76e9',
    pickedTermCloseIcon: 'pickedTermCloseIcon_bbbc76e9',
    errorMessage: 'errorMessage_bbbc76e9',
    errorIcon: 'errorIcon_bbbc76e9',
};
exports.default = styles;
/* tslint:enable */
