// @pnp/spfx-controls-react/lib/controls/filePicker/RecentFilesTab/RecentFilesTab.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./RecentFilesTab.module.css');
var styles = {
    filePicker: 'filePicker_3fa66c68',
    tabsContainer: 'tabsContainer_3fa66c68',
    tabContainer: 'tabContainer_3fa66c68',
    nav: 'nav_3fa66c68',
    tabHeaderContainer: 'tabHeaderContainer_3fa66c68',
    breadcrumbNav: 'breadcrumbNav_3fa66c68',
    breadcrumbNavItem: 'breadcrumbNavItem_3fa66c68',
    tab: 'tab_3fa66c68',
    tabFiles: 'tabFiles_3fa66c68',
    scrollablePaneWrapper: 'scrollablePaneWrapper_3fa66c68',
    actionButtonsContainer: 'actionButtonsContainer_3fa66c68',
    actionButtons: 'actionButtons_3fa66c68',
    actionButton: 'actionButton_3fa66c68',
    tabHeader: 'tabHeader_3fa66c68',
    selectedTab: 'selectedTab_3fa66c68',
    header: 'header_3fa66c68',
    xlg: 'xlg_3fa66c68',
    xxlg: 'xxlg_3fa66c68',
    xxxlg: 'xxxlg_3fa66c68',
    xxxxlg: 'xxxxlg_3fa66c68',
    itemPickerTopBar: 'itemPickerTopBar_3fa66c68',
    commandBarNoChevron: 'commandBarNoChevron_3fa66c68',
    fileItem: 'fileItem_3fa66c68',
    folderItem: 'folderItem_3fa66c68',
    fileTypeIcon: 'fileTypeIcon_3fa66c68',
    fileTypeIconIcon: 'fileTypeIconIcon_3fa66c68',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_3fa66c68',
    iconColumnHeader: 'iconColumnHeader_3fa66c68',
    fileRow: 'fileRow_3fa66c68',
    folderRow: 'folderRow_3fa66c68',
    emptyFolder: 'emptyFolder_3fa66c68',
    emptyFolderImage: 'emptyFolderImage_3fa66c68',
    emptyFolderImageTag: 'emptyFolderImageTag_3fa66c68',
    emptyFolderTitle: 'emptyFolderTitle_3fa66c68',
    emptyFolderSubText: 'emptyFolderSubText_3fa66c68',
    emptyFolderMobile: 'emptyFolderMobile_3fa66c68',
    emptyFolderPc: 'emptyFolderPc_3fa66c68',
    emptyFolderIcon: 'emptyFolderIcon_3fa66c68',
    gridListCell: 'gridListCell_3fa66c68',
    cellRight: 'cellRight_3fa66c68',
    cellLastRow: 'cellLastRow_3fa66c68',
    itemTile: 'itemTile_3fa66c68',
    isFile: 'isFile_3fa66c68',
    isPhoto: 'isPhoto_3fa66c68',
    isVideo: 'isVideo_3fa66c68',
    hasThumbnail: 'hasThumbnail_3fa66c68',
    itemTileContent: 'itemTileContent_3fa66c68',
    itemTileDiscoverFileContainer: 'itemTileDiscoverFileContainer_3fa66c68',
    itemTileFileContainer: 'itemTileFileContainer_3fa66c68',
    itemTileFile: 'itemTileFile_3fa66c68',
    itemTileThumbnail: 'itemTileThumbnail_3fa66c68',
    image: 'image_3fa66c68',
    itemTileCheckCircle: 'itemTileCheckCircle_3fa66c68',
    isSelected: 'isSelected_3fa66c68',
    itemTileNamePlate: 'itemTileNamePlate_3fa66c68',
    itemTileName: 'itemTileName_3fa66c68',
    itemTileChildCount: 'itemTileChildCount_3fa66c68',
    itemTileSubText: 'itemTileSubText_3fa66c68',
    recentGridList: 'recentGridList_3fa66c68',
};
exports.default = styles;
/* tslint:enable */
